avocado/utils/asset.py

```diff
@@ -199,127 +299,105 @@ def fetch(self):
-
-        # If we get to this point, we have to download it from a location.
-        # A writable cache directory is then needed. The first available
-        # writable cache directory will be used.
+        asset_path = self.find_asset_file()
+        # If asset file exists locally, let's create the metadata and return
+        # the path to the file.
+        if asset_path:
+            if self.metadata:
+                self._create_metadata_file(asset_path)
+            return asset_path
+
+        # If we get to this point, we have to download the asset.
+        # We need a writable cache directory. The first available writable
+        # cache directory is used.
         cache_dir = self._get_writable_cache_dir()
-        # Now we have a writable cache_dir. Let's get the asset.
-        # Adding the user defined locations to the urls list:
-        if self.locations is not None:
-            for item in self.locations:
-                urls.append(item)
-
-        cache_relative_dir = self._get_relative_dir(parsed_url)
-        for url in urls:
+
+        # We have a writable cache_dir. Let's get the asset.
+        # Download asset from first available url and return its local path.
+        for location in self.locations:
+            url = os.path.join(location, self.basename)
             urlobj = urlparse.urlparse(url)
+            # Select the correct transport protocol.
             if urlobj.scheme in ['http', 'https', 'ftp']:
-                fetch = self._download
+                fetch_func = self._download
             elif urlobj.scheme == 'file':
-                fetch = self._get_local_file
+                fetch_func = self._get_local_file
             else:
                 raise UnsupportedProtocolError("Unsupported protocol"
                                                ": %s" % urlobj.scheme)
-            asset_file = os.path.join(cache_dir, cache_relative_dir, basename)
-            dirname = os.path.dirname(asset_file)
+
+            # Create the target directory for the asset
+            relative_dir = self._get_relative_dir(location)
+            asset_path = os.path.join(cache_dir, relative_dir, self.basename)
+            dirname = os.path.dirname(asset_path)
             if not os.path.isdir(dirname):
                 os.makedirs(dirname)
+
+            # Let's try to download the asset.
             try:
-                if fetch(urlobj, asset_file):
-                    if self.metadata is not None:
-                        self._create_metadata_file(asset_file)
-                    return asset_file
+                if fetch_func(urlobj, asset_path):
+                    # Success! Create metadata.
+                    if self.metadata:
+                        self._create_metadata_file(asset_path)
+                    return asset_path
+                # If fetch_func returns False, we have a hash mismatch
+                raise EnvironmentError("Hash mismatch for asset %s."
```

Comment:
This exception will be caught by next `except` clause. Example:
```
>>> try:
...     raise EnvironmentError("Hash mismatch for asset")
... except Exception as e:
...     print('%s - %s' % (e.__class__.__name__, str(e)))
... 
OSError - Hash mismatch for asset
```
So the exception never gets to the caller unless you re-throw inside the `except` block.